### PR TITLE
boards: renesas: Add support for EK-RA8D1, MCK_RA8T1

### PR DESCRIPTION
**Pull request by duynguyenxa**
This PR add support for EK-RA8D1 and MCK-RA8T1 based on the EK-RA8M1 development
Note: This should be merge after the https://github.com/zephyrproject-rtos/zephyr/pull/76312

### CODE REVIEW

**Comment by soburi** [comment]
Please mark it as `draft` by the github feature.
スクリーンショット 2024-07-02 061947

**Comment by soburi** [comment]
スクリーンショット 2024-07-02 061947
Please mark it as `draft` by the github feature.

**Comment by soburi** [comment]
スクリーンショット 2024-07-02 062558
Please mark it as `draft` by the github feature.

**Comment by soburi** [comment]
And I think this PR is too huge.
Please split into each board addition and drivers addition.
It is better to add drivers first with it into RA8M1.

**Comment by KhiemNguyenT** [comment]
> And I think this PR is too huge. Please split into each board addition and drivers addition. It is better to add drivers first with it into RA8M1.

@soburi this suggestion will be considered in next rebase. Perhaps, you will give us some comments on these patches, please.

**Comment by KhiemNguyenT** [comment]
I have checked the patches. It's aligned with previous PR. Hope to receive some comments from others.

File: drivers/spi/spi_b_renesas_ra8.c [hunk 1]
```diff
@@ -0,0 +1,711 @@
...
+
+	return 0;
+}
+
+static const struct spi_driver_api ra_spi_driver_api = {.transceive = ra_spi_b_transceive,
+#ifdef CONFIG_SPI_ASYNC
+							.transceive_async =
+								ra_spi_b_transceive_async,
+#endif /* CONFIG_SPI_ASYNC */
+							.release = ra_spi_b_release};
```
Comment (ydamigos) [inline drivers/spi/spi_b_renesas_ra8.c:384]:
```suggestion
static const struct spi_driver_api ra_spi_driver_api = {
	.transceive = ra_spi_b_transceive,
#ifdef CONFIG_SPI_ASYNC
	.transceive_async = ra_spi_b_transceive_async,
#endif /* CONFIG_SPI_ASYNC */
	.release = ra_spi_b_release
};
```

File: drivers/spi/spi_b_renesas_ra8.c [hunk 2]
```diff
@@ -0,0 +1,711 @@
...
+struct ra_spi_data {
+	struct spi_context ctx;
+	uint8_t dfs;
+	struct st_spi_b_instance_ctrl spi;
+	struct st_spi_cfg fsp_config;
+	struct st_spi_b_extended_cfg fsp_config_extend;
+#if CONFIG_SPI_B_INTERRUPT
+	uint32_t data_len;
+#endif
+#if defined(CONFIG_SPI_B_RA_DTC)
```
Comment (ydamigos) [inline drivers/spi/spi_b_renesas_ra8.c:41]: Is it possible for someone to use DTC only on RX or TX?
If yes, we should use tx-dtc and rx-dtc DT properties.
If no, should we consider use only one property for DTC in DT (e.g. use_dtc)?

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: tests/drivers/i2c/i2c_api/src/test_i2c.c [hunk 3]
```diff
@@ -28,6 +28,16 @@
 
 uint32_t i2c_cfg = I2C_SPEED_SET(I2C_SPEED_STANDARD) | I2C_MODE_CONTROLLER;
 
+#define HMC_ADDR (0x1E)
```
Comment (soburi) [inline tests/drivers/i2c/i2c_api/src/test_i2c.c:31]: Please define in Kconfig and use `extra_configs` in testcase.yaml.

**Review by soburi — CHANGES_REQUESTED** [review]

File: soc/renesas/ra/ra8m1/soc.c [hunk 4]
```diff
@@ -17,10 +17,14 @@
 #include <zephyr/arch/arm/nmi.h>
 #include <zephyr/irq.h>
 #include <zephyr/logging/log.h>
+#include <zephyr/sys/barrier.h>
+
 LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
 
 #include <bsp_api.h>
 
+#define CCR_CACHE_ENABLE (0x000E0201)
```
Comment (soburi) [inline soc/renesas/ra/ra8m1/soc.c:26]: Can this value be expressed in the cmsis definition?

Comment (duynguyenxa) [reply soc/renesas/ra/ra8m1/soc.c]: @soburi , I changed it into
`/* Enable Instrcution cache, Branch prediction, Loop and branch info cache */
#define CCR_CACHE_ENABLE (SCB_CCR_IC_Msk | SCB_CCR_BP_Msk | SCB_CCR_LOB_Msk)`

File: drivers/clock_control/clock_control_renesas_ra_cgc.c [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/kernel.h>
+#include <soc.h>
+#include <zephyr/dt-bindings/clock/ra_clock.h>
+#include <zephyr/drivers/clock_control/renesas_ra_cgc.h>
+
+static int clock_control_renesas_ra_on(const struct device *dev, clock_control_subsys_t sys)
+{
+	struct clock_control_ra_subsys_cfg *subsys_clk = (struct clock_control_ra_subsys_cfg *)sys;
```
Comment (soburi) [inline drivers/clock_control/clock_control_renesas_ra_cgc.c:17]: Please do not cast `clock_control_subsys`_t to different types ​​for on, off, and get_rate. For places where the same type is used as an argument in the clock_control API, please use a consistent type according to the definition of the original API.

**Review by soburi — CHANGES_REQUESTED** [review]

File: drivers/spi/Kconfig.renesas_ra8 [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+if SPI_RENESAS_RA8
+
+config SPI_B_INTERRUPT
+	bool "RA MCU SPI B Interrupt Support"
+	help
+	  Enable Interrupt support for the SPI B Driver of RA family.
+
+config SPI_B_RA_DTC
+	bool "RA MCU SPI DTC Support"
+	select USE_RA_FSP_DTC
```
[the remark this answers is not in the document]
Comment (duynguyenxa) [reply drivers/spi/Kconfig.renesas_ra8:23]: @ydamigos, As for now it's seemed there is no good way to get the property value of rx-dtc and tx-dtc without the node name so I will set this enable as default.

File: drivers/spi/spi_b_renesas_ra8.c (hunk 2 above)
Comment (duynguyenxa) [reply drivers/spi/spi_b_renesas_ra8.c:41]: @ydamigos , User are able to to use only tx dtc or rx dtc.
`#ifndef CONFIG_SPI_B_RA_DTC
#define RA_SPI_B_DTC_STRUCT_INIT(index)
#define RA_SPI_B_DTC_INIT(index)
#else
#define RA_SPI_B_DTC_INIT(index)                                                                   \
	do {                                                                                       \
		if (DT_INST_PROP_OR(index, rx_dtc, false)) {                                       \
			ra_spi_data_##index.fsp_config.p_transfer_rx =                             \
				&ra_spi_data_##index.rx_transfer;                                  \
		}                                                                                  \
		if (DT_INST_PROP_OR(index, tx_dtc, false)) {                                       \
			ra_spi_data_##index.fsp_config.p_transfer_tx =                             \
				&ra_spi_data_##index.tx_transfer;                                  \
		}                                                                                  \
	} while (0)`

File: tests/drivers/i2c/i2c_api/src/test_i2c.c (hunk 3 above)
Comment (duynguyenxa) [reply tests/drivers/i2c/i2c_api/src/test_i2c.c]: @soburi , It's updated please have a look

File: drivers/clock_control/clock_control_renesas_ra_cgc.c (hunk 5 above)
Comment (duynguyenxa) [reply drivers/clock_control/clock_control_renesas_ra_cgc.c:17]: @soburi , I refer to some other clock control driver, they also do casting for subsys clock.

e.g. https://github.com/zephyrproject-rtos/zephyr/blob/fd0e46506fabd4c0c91ecb336a39cb6a6aa2b270/drivers/clock_control/clock_stm32_ll_u5.c#L151C9-L151C22

Please correct me if I'm wrong, as my understanding in the description in clock_control.h, the clock_control_subsys_t is designed to be casted to the data structure of sub-system driver instance being used.

/**
 * clock_control_subsys_t is a type to identify a clock controller sub-system.
 * Such data pointed is opaque and relevant only to the clock controller
 * driver instance being used.
 */
typedef void *clock_control_subsys_t;

Comment (soburi) [reply drivers/clock_control/clock_control_renesas_ra_cgc.c:17]: Sorry, I checked something incorrectly, so I'll comment again.

Here, casting itself is not a problem.
The type is defined as void*, so it can be any type.

By the way, how is the clock_control_subsys_t argument processed in get_rate?

Here, the function is expected to be called with the same clock_control_subsys_t value as in the on/off case, so the corresponding clock frequency is expected to be obtained.

In other words, if the same value as a clock_control_subsys_t that allows on and off to operate normally is passed to get_rate, that frequency is expected to be obtained for this. We think it should be possible to get the pclk[abc...] value corresponding to each peripheral, not the system clock.

Comment (duynguyenxa) [reply drivers/clock_control/clock_control_renesas_ra_cgc.c:17]: @soburi -san, this implement treat each pclk as a clock device, hence it will get the rate of pclk[abc...] by the device config. Similar as the implementation of fixed clock driver in Zephyr

Comment (soburi) [reply drivers/clock_control/clock_control_renesas_ra_cgc.c:17]: In Zephyr, it is expected to call the `get_rate` of clock_controller with the value of `clocks` for each node set in the device tree to obtain the clock frequency required by that peripheral. The API of clock_controller is designed with this usage in mind.

In the case of Renesas RA, the FSP mechanism can be used to obtain the clock information required by the driver without using this method,  so this is not a problem at a glance, but this is the FSP way and not Zephyr's. Since clocks affect the entire system, they should be designed appropriately according to Zephyr's custom; the get_rate of clock_controller should be implemented appropriately.

For example, the SCI uses pclka in RA4M1 and the pclkb in RA2A1.
I hope it can switch between these in the device tree description and obtain the frequency of each clock through get_rate.

In practice, I think adding one cell to specify the type of clock is enough.

**Comment by duynguyenxa** [comment]
@soburi ,  As you requested earlier, we separate this PR into smaller chunk of driver support for RA8 series, this PR is now only to add RA8D1 and RA8T1 support based on the existed RA8M1.

File: boards/renesas/ek_ra8d1/ek_ra8d1.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+identifier: ek_ra8d1
+name: Renesas EK-RA8D1
+type: mcu
+arch: arm
+ram: 1024
+flash: 2048
+toolchain:
+  - gnuarmemb
```
Comment (ydamigos) [inline boards/renesas/ek_ra8d1/ek_ra8d1.yaml:8]: we could also add zephyr here, couldn't we?

Comment (duynguyenxa) [reply boards/renesas/ek_ra8d1/ek_ra8d1.yaml:9]: @ydamigos , I added it

**Review by soburi — DISMISSED** [review]
LGTM

File: west.yml [hunk 8]
```diff
@@ -214,7 +214,7 @@ manifest:
         - hal
     - name: hal_renesas
       path: modules/hal/renesas
-      revision: af77d7cdfeeff290593e7e99f54f0c1e2a3f91e6
+      revision: pull/24/head
```
Comment (ydamigos) [inline west.yml:217]: PR was merged in hal_renesas. Please use latest hash.

**Review by ydamigos — CHANGES_REQUESTED** [review]
west.yml needs to be updated.

**Review by ydamigos — DISMISSED** [review]

**Review by ydamigos — APPROVED** [review]

**Review by soburi — APPROVED** [review]

**Review by KhiemNguyenT — APPROVED** [review]
